tokio-postgres/src/copy_both.rs

```diff
@@ -0,0 +1,347 @@
+    CopyComplete,
+    /// We have received the final CommandComplete message for the statement
+    CommandComplete,
+}
+
+/// A CopyBothReceiver is responsible for handling the CopyBoth subprotocol. It ensures that no
+/// matter what the users do with their CopyBothDuplex handle we're always going to send the
+/// correct messages to the backend in order to restore the connection into a usable state.
+///
+/// ```ignore
+///                                          |
+///          <tokio_postgres owned>          |    <userland owned>
+///                                          |
+///  pg -> Connection -> CopyBothReceiver ---+---> CopyBothDuplex
+///                                          |          ^   \
+///                                          |         /     v
+///                                          |      Sink    Stream
+/// ```
+pub struct CopyBothReceiver {
+    /// Receiver of backend messages from the underlying [Connection](crate::Connection)
+    message_receiver: mpsc::Receiver<Result<Message, Error>>,
+    /// Receiver of frontend messages sent by the user using <CopyBothDuplex as Sink>
+    sink_receiver: mpsc::Receiver<FrontendMessage>,
+    /// Sender of CopyData contents to be consumed by the user using <CopyBothDuplex as Stream>
+    stream_sender: mpsc::Sender<Result<Bytes, Error>>,
+    /// The current state of the subprotocol
+    state: CopyBothState,
+    /// Holds a buffered message until we are ready to send it to the user's stream
+    buffered_message: Option<Result<Bytes, Error>>,
+}
+
+impl CopyBothReceiver {
+    pub(crate) fn new(
+        message_receiver: mpsc::Receiver<Result<Message, Error>>,
+        sink_receiver: mpsc::Receiver<FrontendMessage>,
+        stream_sender: mpsc::Sender<Result<Bytes, Error>>,
+    ) -> CopyBothReceiver {
+        CopyBothReceiver {
+            message_receiver,
+            sink_receiver,
+            stream_sender,
+            state: CopyBothState::CopyBoth,
+            buffered_message: None,
+        }
+    }
+
+    /// Convenience method to set the subprotocol into an unexpected message state
+    fn unexpected_message(&mut self) {
+        self.sink_receiver.close();
+        self.buffered_message = Some(Err(Error::unexpected_message()));
+        self.state = CopyBothState::CommandComplete;
+    }
+
+    /// Processes messages from the backend, it will resolve once all backend messages have been
+    /// processed
+    fn poll_backend(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        use CopyBothState::*;
+
+        loop {
+            // Deliver the buffered message (if any) to the user to ensure we can potentially
+            // buffer a new one in response to a server message
+            if self.buffered_message.is_some() {
+                // If the receiver is gone we'll just drop the message
+                let _ = ready!(self.stream_sender.poll_ready(cx));
+                let message = self.buffered_message.take().unwrap();
```

Comment:
Perhaps use an `if ... let` here to avoid the `unwrap()`?

Reply:
Hm, but if `buffered_message` is None then it is definitely a bug in the library because the outer if checked that it's Some(_). I'll replace `unwrap()` with `expect()` to provide an explanation in case this happens in the future